File: 410-split-array-largest-sum/split-array-largest-sum.ts
function splitArray(nums: number[], k: number): number {
    let l = Math.max(...nums);
    let r = nums.reduce((a, b) => a + b, 0);
    let result = r;

    const countPartition = (maxSum: number): number => {
        let partition = 1;
        let curSum = 0;
        for (let i = 0; i < nums.length; i++) {
            if (curSum + nums[i] > maxSum) {
                partition++;
                curSum = nums[i];
            } else {
                curSum += nums[i];
            }
        }
        return partition;
    };

    while (l <= r) {
        let m = Math.floor((l + r) / 2);
        let partition = countPartition(m);
        if (partition <= k) {
            result = m;
            r = m - 1;
        } else {
            l = m + 1;
        }
    }

    return result;
}

File: 410-split-array-largest-sum/split-array-largest-sum.java
class Solution {
    public int splitArray(int[] nums, int k) {
        int ans = 0;
        int l = 0;
        int r = 0;
        for(int num : nums) {
            l = Math.max(num , l);
            r += num;
        }

        while(l <= r) {
            int m = l + (r - l) / 2;
            int partitions = getPartitions(nums ,m);
            if(partitions > k) l = m + 1;
            else {
                ans = m;
                r = m - 1;
            }
        }
        return ans;
    }
    private int getPartitions(int[] nums , int maxSum) {
        int partitions = 1;
        int curSum = 0;
        for(int num : nums) {
            if(curSum + num > maxSum) {
                partitions ++;
                curSum = num;
            }
            else {
                curSum += num;
            }
        }
        return partitions;
    }
}